docs/proposals/20200329-join.md

```diff
@@ -0,0 +1,48 @@
+# KubeOne Join
+
+**Author:** Mohamed Elsayed ([@moelsayed](https://github.com/moelsayed))  
+**Status:** **Draft** | Review | Final
+**Created:** 2020-03-29
+**Last updated:** 2020-03-29
+
+## Abstract
+KubeOne initially supported deploying control plane nodes only, with the option to deploy machine-controller and dynamically provision worker nodes on a configured cloud provider.
+
+PR [#834](https://github.com/kubermatic/kubeone/pull/834) adds support for configuring static worker nodes to be deployed during initial cluster provisioning.
+
+The `join` command should provide allow users to easily add additional worker nodes to an existing KubeOne cluster.
+
+Currently, this is only doable by applying several manual steps on each worker node to deploy the required packages/files and execute `kubeadm join`.
+
+This new subcommand should allow users to add new static worker nodes to the clusters with a single command, and include these nodes in upgrade process as well.
+
+## Goals
+* Implement the `join` subcommand.
+  * Add the ability to join a static worker node to an existing KubeOne cluster.
+  * Support upgrade process for the added static nodes. These nodes will _not_ be in the cluster configuration file and this should be handled correctly.
+
+## Non-goals
+* Support adding control plane nodes.
+* Support node removal/cleanup. This can be done manually by the user.
+
+
+## Implementation
+* Require standard subcommand arguments
+* Require new node configurations
+  * minimal required CLI options to join the nodes
+  * `HostConfig` YAML manifest for the node
+* Parse options and node configuration, create state
+  * Apply defaults on the node
+
+### Join
+* Continuing after the initial steps
+* Run the following tasks on the new node:
+  * `installPrerequisitesOnNode`
+  * `generateKubeadmOnNode`
+  * `joinStaticWorkerInternal`
+* Store node configuration in a ConfigMap/machine object upon successful join.
```

Comment:
Not very happy about this. State shouldn't be stored on the cluster. So, open to ideas!

Reply:
I'm also not happy about this at all. This feature should work like we handle control planes. Have a YAML manifest where you would store the config needed to provision those nodes. Then, it's the operator's responsibility to store and keep that file.